### PR TITLE
Race condition AutoSave fixes

### LINKED ISSUE #1955: race condition in auto save
When a user is performing a task that takes "some time" on the server - and when an autosave happens, this causes a race condition. The likelyhood of this being an issue is a function of the time it takes on the server and the frequency of the autosave.

This manifests itself as a number of errors in the logs:
- "Oops! There was a problem uploading your file"
-  "Any exception during `uploadManuscriptFile`"

Caused by not taking a lock on the manuscript object for updating. Also, should the autosave actually happen?

### PR DESCRIPTION
**Pull request by erezmus**
- [x] Prevent files data being updated on auto save (server side)
- [x] Prevent manuscript data being updated on file upload (server side)
- [x] Prevent AutoSave from triggering when uploading files (client side)

Closes #1955

### CODE REVIEW

**Comment by erezmus** [comment]
AutoSave change needs test

**Comment by will-byrne** [comment]
We have changed the tests to match the new code behaviour (no longer deleting relations not sent back as part of the query).

**Review by will-byrne — APPROVED** [review]

File: packages/component-submission/server/resolvers/removeUploadedManuscript.js [hunk 1]
```diff
@@ -1,29 +1,38 @@
...
+  if (fileModels.length > 0) {
+    await fileModels[0].delete()
+
     logger.debug(`Manuscript file removed`, {
       manuscriptId: vars.id,
       userId: userUuid,
     })
   }
 
+  //   const manuscriptUploadIndex = manuscriptModel.files.findIndex(
```
Comment (diversemix) [inline packages/component-submission/server/resolvers/removeUploadedManuscript.js]: Commented out code!

File: packages/component-submission/server/resolvers/removeUploadedManuscript.js [hunk 2]
```diff
@@ -1,29 +1,38 @@
...
+// .model
+const FileModel = require('@elifesciences/component-model-file').model
 const logger = require('@pubsweet/logger')
 const { S3Storage } = require('@elifesciences/component-service-s3')
 const { Manuscript } = require('../use-cases')
 
 async function removeUploadedManuscript(_, vars, { user }) {
   const userUuid = await User.getUuidForProfile(user)
-  const manuscriptModel = await ManuscriptModel.find(vars.id, userUuid)
+  // const manuscriptModel = await ManuscriptModel.find(vars.id, userUuid)
```
Comment (diversemix) [inline packages/component-submission/server/resolvers/removeUploadedManuscript.js]: Commented out code - please delete

File: packages/component-submission/server/resolvers/removeUploadedManuscript.js [hunk 3]
```diff
@@ -1,29 +1,38 @@
...
-const ManuscriptModel = require('@elifesciences/component-model-manuscript')
-  .model
+// const ManuscriptModel = require('@elifesciences/component-model-manuscript')
+// .model
+const FileModel = require('@elifesciences/component-model-file').model
 const logger = require('@pubsweet/logger')
 const { S3Storage } = require('@elifesciences/component-service-s3')
 const { Manuscript } = require('../use-cases')
 
 async function removeUploadedManuscript(_, vars, { user }) {
```
Comment (diversemix) [inline packages/component-submission/server/resolvers/removeUploadedManuscript.js:8]: Maybe we should rename to `removeUploadedManuscriptFile` ? I
 guess the file too ... to help indicate that it only operates on the file

File: packages/component-submission/server/use-cases/manuscript.js [hunk 4]
```diff
@@ -77,7 +77,11 @@ class Manuscript {
   }
 
   async update(data) {
-    const manuscript = await ManuscriptModel.find(data.id, this.userId)
+    const manuscript = await ManuscriptModel.find(
```
Comment (diversemix) [inline packages/component-submission/server/use-cases/manuscript.js:81]: Personally, I'd like to see a comment to explain that we are just going to update `teams` as files are handled by the upload/remove functionality.

File: packages/component-model-manuscript/entities/manuscript/index.js [hunk 5]
```diff
@@ -150,7 +150,7 @@ class Manuscript extends BaseModel {
     })
     // todo why do I need to do this?
     await Promise.all(
```
Comment (diversemix) [inline packages/component-model-manuscript/entities/manuscript/index.js:151]: Is this another instances of the wrapping in a `Promise.all` for no real reason?

**Comment by diversemix** [comment]
We need to take a look - https://alfred.elifesciences.org/blue/organizations/jenkins/pull-requests-projects%2Felife-xpub/detail/PR-1964/8/pipeline

```
"level":"info","message":"New Files: 5c92a878-aaf9-4d54-bb91-edbbac5cebfd | MANUSCRIPT_SOURCE","timestamp":"2019-05-01T14:08:13.956Z"}
{"level":"error","message":"Attempt to save Manuscript 9ae9c1d2-a9d2-4f3a-91ec-bfdd71cc88bb with updated=Wed May 01 2019 14:08:08 GMT+0000 (UTC)","timestamp":"2019-05-01T14:08:14.053Z"}
{"level":"error","message":"Error during final update: Error: Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.072Z"}
{"level":"error","message":"uploadManuscriptFile threw this error: Error: Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.072Z"}
{"level":"info","message":"::ffff:172.19.0.7 [01/May/2019:14:08:14 +0000] POST /graphql 200 UpdateManuscript 2176 795.892 ms\n","timestamp":"2019-05-01T14:08:14.091Z"}
{"error":{"message":"Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","locations":[{"line":2,"column":3}],"path":["uploadManuscript"]},"level":"error","message":"Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.194Z"}
```

File: packages/component-model-manuscript/entities/manuscript/index.js (hunk 5 above)
Comment (erezmus) [reply packages/component-model-manuscript/entities/manuscript/index.js:151]: the reason is that if we used a forEach, then it wouldn't wait until all the $loadRelated calls are done

https://medium.com/@pyapar/another-approach-using-map-instead-of-foreach-82a85f536c9c

Comment (diversemix) [reply packages/component-model-manuscript/entities/manuscript/index.js:151]: Maybe let's get rid of the comment then?

**Comment by diversemix** [comment]
> We need to take a look - https://alfred.elifesciences.org/blue/organizations/jenkins/pull-requests-projects%2Felife-xpub/detail/PR-1964/8/pipeline
> 
> ```
> "level":"info","message":"New Files: 5c92a878-aaf9-4d54-bb91-edbbac5cebfd | MANUSCRIPT_SOURCE","timestamp":"2019-05-01T14:08:13.956Z"}
> {"level":"error","message":"Attempt to save Manuscript 9ae9c1d2-a9d2-4f3a-91ec-bfdd71cc88bb with updated=Wed May 01 2019 14:08:08 GMT+0000 (UTC)","timestamp":"2019-05-01T14:08:14.053Z"}
> {"level":"error","message":"Error during final update: Error: Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.072Z"}
> {"level":"error","message":"uploadManuscriptFile threw this error: Error: Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.072Z"}
> {"level":"info","message":"::ffff:172.19.0.7 [01/May/2019:14:08:14 +0000] POST /graphql 200 UpdateManuscript 2176 795.892 ms\n","timestamp":"2019-05-01T14:08:14.091Z"}
> {"error":{"message":"Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","locations":[{"line":2,"column":3}],"path":["uploadManuscript"]},"level":"error","message":"Data Integrity Error property updated set to Wed May 01 2019 14:08:08 GMT+0000 (UTC): is older than the one stored in the database!","timestamp":"2019-05-01T14:08:14.194Z"}
> ```

I've made the commit `3cdd6ea` to try and reduce this occuring. /cc @will-byrne @erezmus

**Comment by erezmus** [comment]
Regarding the output from the log, its trying to save manuscript with updated of `Wed May 01 2019 14:08:08 GMT+0000` at `2019-05-01T14:08:14.053Z`

However with the lines:

> {"level":"info","message":"Manuscript new index 5c92a878-aaf9-4d54-bb91-edbbac5cebfd | dummy-manuscript-2.pdf","timestamp":"2019-05-01T14:08:13.843Z"}
> {"level":"info","message":"New Files: 5c92a878-aaf9-4d54-bb91-edbbac5cebfd | MANUSCRIPT_SOURCE","timestamp":"2019-05-01T14:08:13.956Z"}

The first line above occurs before the ManuscriptModel.find() on line 84 in the helpers/manuscript.js so i would expect if there had been an update that it would get the latest value.

I propose to log the value of the current updated field in the database to see what we are comparing against in `needsRefresh`

@diversemix shall we do that?

File: packages/component-model-manuscript/entities/manuscript/index.js (hunk 5 above)
Comment (diversemix) [reply packages/component-model-manuscript/entities/manuscript/index.js:151]: @erezmus ... can you remove this comment- as I think its misleading if you understand  it now :)

**Review by diversemix — APPROVED** [review]

**Pull request merged**